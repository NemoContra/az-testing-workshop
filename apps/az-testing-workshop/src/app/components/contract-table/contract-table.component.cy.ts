import { registerLocaleData } from '@angular/common';
import localeDE from '@angular/common/locales/de';
import { provideNoopAnimations } from '@angular/platform-browser/animations';
import { NxExpertModule } from '@aposin/ng-aquila/config';

import { ContractTableComponent } from './contract-table.components';
import { MountConfig } from 'cypress/angular';
import { expectCorrectTableRow } from '../../../../cypress/support/helpers.po';
import { provideRouter } from '@angular/router';
import { LOCALE_ID } from '@angular/core';
import { DummyRouterDestinationComponent, mockContracts } from '../../../../../../libs/shared/util/test-helpers/src';

registerLocaleData(localeDE);

describe(ContractTableComponent.name, () => {
  const mountConfig: MountConfig<ContractTableComponent> = {
    imports: [
      NxExpertModule,
      DummyRouterDestinationComponent
    ],
    providers: [provideNoopAnimations(), provideRouter([{
      path: 'details/:id',
      component: DummyRouterDestinationComponent
    }]), { provide: LOCALE_ID, useValue: 'de-DE' }]
  };

  beforeEach(() => {
    cy.viewport(1600, 800);
  });


  it('should show the correct contracts which are provided as input value', () => {
    cy.mount(ContractTableComponent, { ...mountConfig, componentProperties: { contracts: mockContracts } });

    cy.get('div.scroll-container table[nxTable]').should('be.visible');

    cy.get('table thead tr th').should('have.length', 7);
    cy.get('table thead tr th').eq(0).should('have.text', 'Vertragsnummer');
    cy.get('table thead tr th').eq(1).should('have.text', 'Vorname');
    cy.get('table thead tr th').eq(2).should('have.text', 'Nachname');
    cy.get('table thead tr th').eq(3).should('have.text', 'Geburtsdatum');
    cy.get('table thead tr th').eq(4).should('have.text', 'Vertragsbeginn');
    cy.get('table thead tr th').eq(5).should('have.text', 'Vertragsende');
    cy.get('table thead tr th').eq(6).should('have.text', 'Aktion');

    cy.get('table tbody tr').should('have.length', 2);
    cy.get('table tbody tr').eq(0).find('td').should('have.length', 7);
    cy.get('table tbody tr').eq(1).find('td').should('have.length', 7);

    expectCorrectTableRow(0, '1/2345678/9', 'Homer', 'Simpson', '16.05.1961', '01.01.2024', '');
    expectCorrectTableRow(1, '1/2345678/8', 'Bart', 'Simpson', '21.08.1995', '01.02.2024', '');
  });

  it('should have a context-menu with the correct entries', () => {
    cy.mount(ContractTableComponent, { ...mountConfig, componentProperties: { contracts: mockContracts } });

    cy.get('nx-context-menu').should('not.be.visible');

    cy.get('table tbody tr').eq(0).find('td').eq(6).find('button').click();

    cy.get('div.nx-context-menu').should('be.visible');
    cy.get('div.nx-context-menu button').should('have.length', 4);

    cy.get('div.nx-context-menu button').eq(0).should('contain.text', 'Details anzeigen');
    cy.get('div.nx-context-menu button').eq(0).should('have.attr', 'type', 'button');
    cy.get('div.nx-context-menu button').eq(0).should('have.attr', 'nxContextMenuItem');

    cy.get('div.nx-context-menu button').eq(0).click();

    cy.url().should('contain', 'details/123456788');
  });

  it('should show an empty table hint if no contracts are provided', () => {
    cy.mount(ContractTableComponent, { ...mountConfig, componentProperties: { contracts: [] } });

    cy.get('table tbody tr td').should('have.length', 1);
    cy.get('table tbody tr td').should('have.text', 'Keine VertrÃ¤ge vorhanden.')
  });

  it('should emit the correct output event when searching for a contract', () => {
    cy.mount(ContractTableComponent, { ...mountConfig, componentProperties: { contracts: [] }, autoSpyOutputs: true });

    cy.get('nx-formfield input').type('Homer');

    cy.get('@queryChange').should('have.been.calledOnce');
    cy.get('@queryChange').should('have.been.calledWith', 'Homer');
  });
});
