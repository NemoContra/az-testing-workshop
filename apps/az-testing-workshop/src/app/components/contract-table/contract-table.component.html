<nx-formfield>
  <input
    nxInput
    type="text"
    [ngModel]="query()"
    (ngModelChange)="query.set($event)"
  />
  <span nxFormfieldPrefix>
    <nx-icon name="search" size="s"></nx-icon>
  </span>
</nx-formfield>
<div class="scroll-container">
  <table nxTable class="nx-table--scrollable">
    <thead>
      <tr nxTableRow>
        <th nxHeaderCell>Vertragsnummer</th>
        <th nxHeaderCell>Vorname</th>
        <th nxHeaderCell>Nachname</th>
        <th nxHeaderCell>Geburtsdatum</th>
        <th nxHeaderCell>Vertragsbeginn</th>
        <th nxHeaderCell>Vertragsende</th>
        <th nxHeaderCell>Aktion</th>
      </tr>
    </thead>
    <tbody>
      @if (contracts?.length === 0) {
      <tr>
        <td colspan="7">Keine Vertr채ge vorhanden.</td>
      </tr>
      } @for (contract of contracts; track contract.id) {
      <tr>
        <td nxTableCell>{{ contract.contractNumber }}</td>
        <td nxTableCell>{{ contract.person.firstname }}</td>
        <td nxTableCell>{{ contract.person.lastname }}</td>
        <td nxTableCell>{{ contract.person.dateOfBirth | date }}</td>
        <td nxTableCell>{{ contract.start | date }}</td>
        <td nxTableCell>{{ contract.end ? (contract.end | date) : '-' }}</td>
        <td nxTableCell>
          <nx-context-menu #menu>
            <button
              nxContextMenuItem
              type="button"
              [routerLink]="['/details', contract.id]"
            >
              Details anzeigen
            </button>
            <button
              nxContextMenuItem
              type="button"
              [routerLink]="['/transaktion', contract.id]"
            >
              Transaktion durchf체hren
            </button>
            <button
              nxContextMenuItem
              type="button"
              [routerLink]="['/transaktion', contract.id]"
              [queryParams]="aenderungNachnameQueryParams"
            >
              Nachname 채ndern
            </button>
            <button
              nxContextMenuItem
              type="button"
              [routerLink]="['/transaktion', contract.id]"
              [queryParams]="kuendigungQueryParams"
            >
              K체ndigung
            </button>
          </nx-context-menu>
          <button
            nxIconButton="tertiary small"
            [nxContextMenuTriggerFor]="menu"
            aria-label="Open menu"
            type="button"
          >
            <nx-icon aria-hidden="true" name="ellipsis-h"></nx-icon>
          </button>
        </td>
      </tr>
      }
    </tbody>
  </table>
</div>
