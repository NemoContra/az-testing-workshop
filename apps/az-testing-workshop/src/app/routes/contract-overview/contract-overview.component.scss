.spinner-outlet {
  height: 20px;
  display: flex;
  justify-content: flex-end;
}

contract-table {
  display: block;
  margin: 36px;
}
