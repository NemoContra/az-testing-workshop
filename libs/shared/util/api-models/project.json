{
  "name": "shared-util-api-models",
  "$schema": "../../../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/shared/util/api-models/src",
  "projectType": "library",
  "targets": {
    "lint": {
      "executor": "@nx/eslint:lint"
    }
  },
  "tags": ["scope:shared", "type:util"]
}
