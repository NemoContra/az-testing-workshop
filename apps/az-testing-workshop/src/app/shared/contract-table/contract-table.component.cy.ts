import { registerLocaleData } from '@angular/common';
import localeDE from '@angular/common/locales/de';
import { provideNoopAnimations } from '@angular/platform-browser/animations';
import { NxExpertModule } from '@aposin/ng-aquila/config';

import { ContractTableComponent } from './contract-table.components';
import { createOutputSpy, MountConfig } from 'cypress/angular';
import { expectCorrectTableRow } from '../../../../cypress/support/helpers.po';
import { provideRouter } from '@angular/router';
import { LOCALE_ID } from '@angular/core';
import { mockContracts } from '@az-testing-workshop/shared/util/mock-data';
import { mockComponent } from '@az-testing-workshop/shared/util/test-helpers';

registerLocaleData(localeDE);

describe(ContractTableComponent.name, () => {
  const mountConfig: MountConfig<ContractTableComponent> = {
    imports: [NxExpertModule, ContractTableComponent],
    providers: [
      provideNoopAnimations(),
      provideRouter([
        {
          path: '**',
          component: mockComponent({ selector: 'routing-dummy' }),
        },
      ]),
      { provide: LOCALE_ID, useValue: 'de-DE' },
    ],
  };

  beforeEach(() => {
    cy.viewport(1600, 800);
  });

  it('should show the correct contracts which are provided as input value', () => {
    cy.mount(ContractTableComponent, {
      ...mountConfig,
      componentProperties: { contracts: mockContracts },
    });
  });

  it('should have a context-menu with the correct entries', () => {
    cy.mount(ContractTableComponent, {
      ...mountConfig,
      componentProperties: { contracts: mockContracts },
    });
  });

  it('should show an empty table hint if no contracts are provided', () => {
    cy.mount(ContractTableComponent, {
      ...mountConfig,
      componentProperties: { contracts: [] },
    });
  });

  it('should emit the correct output event when searching for a contract', async () => {
    cy.mount<ContractTableComponent>(
      '<contract-table [contracts]="contracts" (queryChange)="queryChange.emit($event)"></contract-table>',
      {
        ...mountConfig,
        componentProperties: {
          contracts: mockContracts,
          queryChange: createOutputSpy<string>('queryChangeSpy'),
        },
      }
    );

    cy.get('nx-formfield input').type('Homer');
  });
});
