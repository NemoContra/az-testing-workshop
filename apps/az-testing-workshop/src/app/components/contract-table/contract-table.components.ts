import {
  ChangeDetectionStrategy,
  Component,
  Input,
  Output,
  signal,
} from '@angular/core';
import { Contract } from '@az-testing-workshop/shared/util/api-models';
import { DatePipe } from '@angular/common';
import { NxTableModule } from '@aposin/ng-aquila/table';
import { NxLinkModule } from '@aposin/ng-aquila/link';
import { RouterLink } from '@angular/router';
import { NxInputModule } from '@aposin/ng-aquila/input';
import { NxIconModule } from '@aposin/ng-aquila/icon';
import { FormsModule } from '@angular/forms';
import { toObservable } from '@angular/core/rxjs-interop';
import { NxMenuModule } from '@aposin/ng-aquila/menu';
import { NxButtonModule } from '@aposin/ng-aquila/button';
import { NxContextMenuModule } from '@aposin/ng-aquila/context-menu';
import { transactionQueryParam } from '../../common/transaction-type';

@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  selector: 'contract-table',
  standalone: true,
  templateUrl: './contract-table.component.html',
  imports: [
    NxTableModule,
    NxLinkModule,
    RouterLink,
    DatePipe,
    NxInputModule,
    NxIconModule,
    FormsModule,
    NxMenuModule,
    NxButtonModule,
    NxContextMenuModule,
  ],
})
export class ContractTableComponent {
  @Input({ required: true }) contracts?: Contract[];
  query = signal('');
  @Output() queryChange = toObservable(this.query);

  aenderungNachnameQueryParams = transactionQueryParam('AenderungNachname');
  kuendigungQueryParams = transactionQueryParam('Kuendigung');
}
