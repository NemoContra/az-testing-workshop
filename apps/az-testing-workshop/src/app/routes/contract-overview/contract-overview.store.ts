import {
  patchState,
  signalStore,
  withHooks,
  withMethods,
  withState,
} from '@ngrx/signals';
import { Contract } from '@az-testing-workshop/shared/util/api-models';
import { rxMethod } from '@ngrx/signals/rxjs-interop';
import { pipe, switchMap, tap } from 'rxjs';
import { inject } from '@angular/core';
import { HttpErrorResponse } from '@angular/common/http';
import { tapResponse } from '@ngrx/operators';
import { ContractService } from '../../services/contract.service';

export type ContractOverviewState = {
  contracts: Contract[] | undefined;
  errorCode: number | undefined;
  loading: boolean;
  query: string;
};

export const initialContractOverviewState: ContractOverviewState = {
  contracts: undefined,
  errorCode: undefined,
  loading: false,
  query: '',
};

export const ContractOverviewStore = signalStore(
  { providedIn: 'root' },
  withState(initialContractOverviewState),
  withMethods((store, contractService = inject(ContractService)) => ({
    setQuery: (query: string) => patchState(store, { query }),
    getContracts: rxMethod<string>(
      pipe(
        tap(() => patchState(store, { loading: true })),
        switchMap((query) =>
          contractService.getContracts(query).pipe(
            tapResponse({
              next: (contracts) => patchState(store, { contracts }),
              error: ({ status }: HttpErrorResponse) =>
                patchState(store, { errorCode: status }),
              finalize: () => patchState(store, { loading: false }),
            })
          )
        )
      )
    ),
  })),
  withHooks(({ getContracts, query }) => ({
    onInit: () => {
      getContracts(query);
    },
  }))
);

export type ContractOverviewStore = InstanceType<typeof ContractOverviewStore>;
