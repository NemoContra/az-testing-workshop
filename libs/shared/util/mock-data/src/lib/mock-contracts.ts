import { Contract } from '@az-testing-workshop/shared/util/api-models';

export const mockContracts: Contract[] = [
  {
    id: '123456789',
    contractNumber: '1/2345678/9',
    start: '2024-01-01',
    premium: 42.42,
    person: {
      firstname: 'Homer',
      lastname: 'Simpson',
      dateOfBirth: '1961-05-16',
    },
  },
  {
    id: '123456788',
    contractNumber: '1/2345678/8',
    start: '2024-02-01',
    premium: 99.0,
    person: {
      firstname: 'Bart',
      lastname: 'Simpson',
      dateOfBirth: '1995-08-21',
    },
  },
];
