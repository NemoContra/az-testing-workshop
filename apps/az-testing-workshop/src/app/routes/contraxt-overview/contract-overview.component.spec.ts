import { createHostFactory, mockProvider } from '@ngneat/spectator/jest';
import ContractOverviewComponent from './contract-overview.component';
import { Router } from '@angular/router';
import { ContractTableComponent } from '../../components/contract-table/contract-table.components';
import { ContractOverviewStore } from './contract-overview.store';
import { signal } from '@angular/core';
import { Contract } from '@az-testing-workshop/shared/util/api-models';
import { mockContracts } from '../../test-helpers/mock-contracts';
import { MockComponent, MockModule } from 'ng-mocks';
import { NxErrorModule } from '@aposin/ng-aquila/base';
import { NxSpinnerModule } from '@aposin/ng-aquila/spinner';

describe('ContractOverview', () => {
  const contractOverviewStoreMock = {
    errorCode: signal(false),
    loading: signal(false),
    contracts: signal<Contract[] | undefined>(undefined),
    setQuery: jest.fn(),
  };

  const createComponent = createHostFactory({
    component: ContractOverviewComponent,
    providers: [
      mockProvider(Router, { navigate: jest.fn().mockResolvedValue(true) }),
      mockProvider(ContractOverviewStore, contractOverviewStoreMock),
    ],
    overrideComponents: [
      [
        ContractOverviewComponent,
        {
          remove: {
            imports: [ContractTableComponent, NxErrorModule, NxSpinnerModule],
          },
          add: {
            imports: [
              MockComponent(ContractTableComponent),
              MockModule(NxErrorModule),
              MockModule(NxSpinnerModule),
            ],
          },
        },
      ],
    ],
  });

  afterEach(() => jest.clearAllMocks());

  it('should render loading-spinner', () => {
    contractOverviewStoreMock.loading.set(true);
    const spectator = createComponent('<contract-overview />');
    expect(spectator.fixture).toMatchSnapshot();
  });

  it('should render error message', () => {
    contractOverviewStoreMock.loading.set(false);
    contractOverviewStoreMock.errorCode.set(true);
    const spectator = createComponent('<contract-overview />');
    expect(spectator.fixture).toMatchSnapshot();
  });

  it('should render contracts-table', () => {
    contractOverviewStoreMock.contracts.set(mockContracts);
    const spectator = createComponent('<contract-overview />');
    expect(spectator.fixture).toMatchSnapshot();
    expect(spectator.query(ContractTableComponent)?.contracts).toEqual(
      mockContracts
    );
  });

  it('should trigger routing when search is used', () => {
    contractOverviewStoreMock.contracts.set(mockContracts);
    const spectator = createComponent('<contract-overview />');
    const router = spectator.inject(Router);
    spectator.triggerEventHandler('contract-table', 'queryChange', 'Bart');
    expect(router.navigate).toHaveBeenCalledTimes(1);
    expect(router.navigate).toHaveBeenCalledWith([], {
      queryParams: {
        query: 'Bart',
      },
    });
  });

  it('should set the queryParam from router', () => {
    contractOverviewStoreMock.contracts.set(mockContracts);
    const spectator = createComponent('<contract-overview />', {
      props: {
        query: 'Bart',
      },
    });
    expect(contractOverviewStoreMock.setQuery).toHaveBeenCalledTimes(1);
    expect(contractOverviewStoreMock.setQuery).toHaveBeenCalledWith('Bart');
    spectator.setInput({ query: 'Simpson' });
    expect(contractOverviewStoreMock.setQuery).toHaveBeenCalledTimes(2);
    expect(contractOverviewStoreMock.setQuery).toHaveBeenNthCalledWith(
      2,
      'Simpson'
    );
  });
});
