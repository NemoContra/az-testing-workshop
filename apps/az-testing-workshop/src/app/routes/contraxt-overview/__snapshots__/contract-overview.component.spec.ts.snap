// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContractOverview should render contracts-table 1`] = `
<lib-ngneat-host-component>
  <contract-overview>
    <div
      class="spinner-outlet"
    >
    </div>
    <contract-table />
  </contract-overview>
</lib-ngneat-host-component>
`;

exports[`ContractOverview should render error message 1`] = `
<lib-ngneat-host-component>
  <contract-overview>
    <nx-error>
      An error occurred
    </nx-error>
    <div
      class="spinner-outlet"
    >
    </div>
  </contract-overview>
</lib-ngneat-host-component>
`;

exports[`ContractOverview should render loading-spinner 1`] = `
<lib-ngneat-host-component>
  <contract-overview>
    <div
      class="spinner-outlet"
    >
      <nx-spinner
        size="medium"
      />
    </div>
  </contract-overview>
</lib-ngneat-host-component>
`;
